In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

import matplotlib.pyplot as plt
import seaborn as sns
import prepare

In [2]:
train, validate, test = prepare.wrangle_data('titanic')

In [3]:
train.head()

survived  pclass     sex        age  sibsp  parch   fare  embark_town  \
428         0       3    male  29.699118      0      0   7.75   Queenstown   
480         0       3    male   9.000000      5      2  46.90  Southampton   
888         0       3  female  29.699118      1      2  23.45  Southampton   
822         0       1    male  38.000000      0      0   0.00  Southampton   
612         1       3  female  29.699118      1      0  15.50   Queenstown   

     alone  sex_male  embark_town_Queenstown  embark_town_Southampton  
428      1         1                       1                        0  
480      0         1                       0                        1  
888      0         0                       0                        1  
822      1         1                       0                        1  
612      0         0                       1                        0

In [4]:
train.survived.value_counts()

0    307
1    191
Name: survived, dtype: int64

In [7]:
baseline = (train.survived == 0).mean()

In [10]:
print(f'our baseline prediction for suvival is death, with an accuracy percentage\
 of: {baseline:.2%}')

our baseline prediction for suvival is death, with an accuracy percentage of: 61.65%


## 1.
Create a model that includes only age, fare, and pclass. Does this model perform better than your baseline?

In [11]:
X_cols = ['age', 'fare', 'pclass']
y_col = 'survived'

In [12]:
X_train = train[X_cols]
y_train = train[y_col]
X_val = validate[X_cols]
y_val = validate[y_col]
X_test = test[X_cols]
y_test = test[y_col]

In [23]:
#create the logistic regression model object
lr_1 = LogisticRegression()

In [24]:
#fit the lr model
lr_1.fit(X_train, y_train)

LogisticRegression()

In [25]:
# predict with the model
lr_1_preds = lr_1.predict(X_train)

In [29]:
# Does this model perform better than your baseline?
print(f'the first logistic regression model score is: {lr_1.score(X_train, y_train):.2%}')

the first logistic regression model score is: 69.68%


The first logistic regression model (default settings on age, fare, and pclass) performed better than the baseline model.

## 2.
Include sex in your model as well. Note that you'll need to encode or create a dummy variable of this feature before including it in a model.

In [20]:
train.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embark_town', 'alone', 'sex_male', 'embark_town_Queenstown',
       'embark_town_Southampton'],
      dtype='object')

In [21]:
X_cols_2 = ['age', 'fare', 'pclass', 'sex_male']

In [22]:
X2_train = train[X_cols_2]

In [27]:
#create the logistic regression model object
lr_2 = LogisticRegression()

#fit the lr model
lr_2.fit(X2_train, y_train)

# predict with the model
lr_2_preds = lr_2.predict(X2_train)

In [30]:
# Does this model perform better than your baseline?
print(f'the second logistic regression model \
score is: {lr_2.score(X2_train, y_train):.2%}')

the second logistic regression model score is: 79.92%


## 3.
Try out other combinations of features and models.

In [31]:
#create the logistic regression model object
lr_3 = LogisticRegression(penalty='none')

#fit the lr model
lr_3.fit(X_train, y_train)

# predict with the model
lr_3_preds = lr_3.predict(X_train)

In [32]:
# Does this model perform better than your baseline?
print(f'the third logistic regression model \
score is: {lr_3.score(X_train, y_train):.2%}')

the third logistic regression model score is: 69.68%


Changing the penalty option to 'none' apears to have done nothing.

In [34]:
#create the logistic regression model object
lr_4 = LogisticRegression(penalty='l2', C=0.2)

#fit the lr model
lr_4.fit(X_train, y_train)

# predict with the model
lr_4_preds = lr_4.predict(X_train)

In [69]:
# Does this model perform better than your baseline?
print(f'the fourth logistic regression model \
score is: {lr_4.score(X_train, y_train):.2%}')

the fourth logistic regression model score is: 70.08%


Changing the C option to 0.2 seems to have improved the score on the training data

In [37]:
#create the logistic regression model object
lr_5 = LogisticRegression(penalty='l2', solver='newton-cg')

#fit the lr model
lr_5.fit(X2_train, y_train)

# predict with the model
lr_5_preds = lr_5.predict(X2_train)

In [68]:
# Does this model perform better than your baseline?
print(f'the fifth logistic regression model \
score is: {lr_5.score(X2_train, y_train):.2%}')

the fifth logistic regression model score is: 79.92%


Using the 'newton-cq' solver method on the second training dataset seems to have much about the same score as using lbfgs

In [41]:
#create the logistic regression model object
lr_6 = LogisticRegression(penalty='l2', solver='newton-cg', C=0.1)

#fit the lr model
lr_6.fit(X2_train, y_train)

# predict with the model
lr_6_preds = lr_6.predict(X2_train)

In [67]:
# Does this model perform better than your baseline?
print(f'the sixth logistic regression model \
score is: {lr_6.score(X2_train, y_train):.2%}')

the sixth logistic regression model score is: 78.51%


lowering the C value to 0.1 using the newton-cg solver seems to have lowered the score.

In [47]:
X_cols_3 = ['age', 'fare', 'pclass', 'sex_male', 'alone']

In [48]:
X3_train = train[X_cols_3]

In [51]:
X3_train.head()

age   fare  pclass  sex_male  alone
428  29.699118   7.75       3         1      1
480   9.000000  46.90       3         1      0
888  29.699118  23.45       3         0      0
822  38.000000   0.00       1         1      1
612  29.699118  15.50       3         0      0

In [52]:
#create the logistic regression model object
lr_7 = LogisticRegression()

#fit the lr model
lr_7.fit(X3_train, y_train)

# predict with the model
lr_7_preds = lr_7.predict(X3_train)

In [66]:
# Does this model perform better than your baseline?
print(f'the seventh logistic regression model \
score is: {lr_7.score(X3_train, y_train):.2%}')

the seventh logistic regression model score is: 79.52%


In [54]:
#create the logistic regression model object
lr_8 = LogisticRegression(penalty='l2', solver='newton-cg', C=0.1)

#fit the lr model
lr_8.fit(X3_train, y_train)

# predict with the model
lr_8_preds = lr_8.predict(X3_train)

In [65]:
# Does this model perform better than your baseline?
print(f'the eighth logistic regression model \
score is: {lr_8.score(X3_train, y_train):.2%}')

the eighth logistic regression model score is: 79.12%


## 4.
Use you best 3 models to predict and evaluate on your validate sample.

In [64]:
# Does this model perform better than your baseline?
print(f'the second logistic regression model \
score is: {lr_2.score(X2_train, y_train):.2%}')
# Does this model perform better than your baseline?
print(f'the fifth logistic regression model \
score is: {lr_5.score(X2_train, y_train):.2%}')
# Does this model perform better than your baseline?
print(f'the seventh logistic regression model \
score is: {lr_7.score(X3_train, y_train):.2%}')

the second logistic regression model score is: 79.92%
the fifth logistic regression model score is: 79.92%
the seventh logistic regression model score is: 79.52%


In [57]:
X2_val = validate[X_cols_2]
X3_val = validate[X_cols_3]

In [63]:
# Does this model perform better than your baseline?
print(f'the second logistic regression model \
score on the validation data is: {lr_2.score(X2_val, y_val):.2%}')
# Does this model perform better than your baseline?
print(f'the fifth logistic regression model \
score on the validation data is: {lr_5.score(X2_val, y_val):.2%}')
# Does this model perform better than your baseline?
print(f'the seventh logistic regression model \
score on the validation data is: {lr_7.score(X3_val, y_val):.2%}')

the second logistic regression model score on the validation data is: 78.04%
the fifth logistic regression model score on the validation data is: 78.04%
the seventh logistic regression model score on the validation data is: 77.57%


Models 2 and 5 (using 'age', 'fare', 'pclass', 'sex_male', with lbfgs / newton-cg solver) seem to have the same level of performance, and both had the best accuracy score.

## 5.
Choose you best model from the validation performation, and evaluate it on the test dataset. How do the performance metrics compare to validate? to train?

In [60]:
X2_test = test[X_cols_2]

In [62]:
print(f'the second logistic regression model \
score on the test data is: {lr_2.score(X2_test, y_test):.2%}')

the second logistic regression model score on the test data is: 81.01%


Model 2 (using 'age', 'fare', 'pclass', 'sex_male', with lbfgs) performed better on the test dataset than either the training data or the validation data. With an accuracy score of 81.01% on the test data, 79.92% on the training data, and 78.04% on the validation data.